File: longest_substring_without_repeat.py
# Time Complexity: O(n) where n is the length of the string s
# Space Complexity: O(1) since the hashmap will have atmost 26 characters
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
We find the longest substring without repeating characters by using a hashmap to store the index of each character within our 
substring as we itereate through s. When iterating through s, if we see that the character is already is in the hashmap, we 
move the first index of our substring to the index of that character in the hashmap and in this way maintaining uniqueness.
"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == None or len(s) == 0:
            return 0
        
        hashmap = dict()
        first_index = 0
        longest_substr = 0
        current_substr = 0

        for index in range(len(s)):
            if s[index] in hashmap:
                if hashmap[s[index]] == first_index:
                    hashmap[s[index]] = index
                    first_index += 1
                else:
                    first_index = hashmap[s[index]] + 1
                    hashmap = {s[i]:i for i in range(first_index, index + 1)}
                    current_substr = index + 1 - first_index

            else:
                hashmap[s[index]] = index
                current_substr += 1
                if current_substr > longest_substr:
                    longest_substr = current_substr


        return longest_substr

File: LongestSubstringwithoutRepeat.java
import java.util.HashMap;

//Time Complexity: O(n)
//Space Complexity: O(1)

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s==null || s.length()==0) return 0;
        
        int slow=0;
        int max=0;
        HashMap<Character,Integer> map=new HashMap<>();
        
        for(int i=0;i<s.length();i++){   //instead of the fast ptr.
            char c=s.charAt(i);
            if(map.containsKey(c)){   //there is a repeat
                slow=Math.max(slow, map.get(c));
            }
            map.put(c,i+1); //HashMap contains character along with its index + 1, for the smoother movement of slow ptr 
            max=Math.max(max, i-slow+1); //max length of substring
        }
        
        return max;
        
        
    }
}